### ISSUE TITLE
about btcv dataset training nextou in nnunetv2

### ISSUE THREAD

**Issue by qiaoqiangPro** [PengchengShi1220/NexToU#9]
hello, bro
    When I tried to reproduce your code, I found an error reported, it seems to be about the number of channels not matching. Also whether I train with 3d_fullres or nextou, I have the log position not found. 
![image](https://github.com/PengchengShi1220/NexToU/assets/80230735/14a311aa-9fe6-4f8a-a3e5-c69907554a7f)

Request for assistance, many thanks!

**Comment by PengchengShi1220** [comment]
Hello,

The issue is likely because the NexToUv1.1 version requires the initial channels to be set as a multiple of 3. Please use the latest [NexToUv1.2](https://github.com/PengchengShi1220/NexToU/releases/download/v1.2.0/NexToU_v1.2_nnU-Net_v2.2.zip) version (based on nnUNet v2.2), which aligns with the default nnUNet settings. The initial channel is set to 32, which is \(2^5\) and generally runs faster than multiples of 3. Additionally, nnUNet v2.2 should resolve the "log position not found" issue. I have tested it locally and it runs smoothly.

Best,
Pengcheng

**Comment by qiaoqiangPro** [comment]
Thanks,
I have downloaded the latest version and it has been resolved. I would like to ask, I see that under nnUNet_preprocessed/Dataset111_Synapse_CT/splits_final.json that I generated, it's 14 training sets and validation sets, is this correct? I went ahead and put the 12 test sets under imagesTs and labelsTs for the final test.
![image](https://github.com/PengchengShi1220/NexToU/assets/80230735/b9a914c3-04d6-4650-a267-0021566c820f)

**Comment by PengchengShi1220** [comment]
If you need to use an 18 training and 12 validation (test) data split, you can manually modify the `nnUNet_preprocessed/Dataset111_Synapse_CT/splits_final.json` file. All 30 cases should be placed in the `imagesTr` directory, with 18 cases for training and 12 cases for validation (which will also serve as the test set). 

For your convenience, I've already uploaded a `splits_final.json` file for the BTCV dataset with fold 0, using the 18 train and 12 val (test) split. You can refer to it here: [splits_final.json](https://github.com/PengchengShi1220/NexToU/blob/1075b40d54a8c0867c83f247b1d29c0dca236c7f/nnUNet_preprocessed/Dataset111_Synapse_CT/splits_final.json#L3).

**Comment by qiaoqiangPro** [comment]
Received Thank you very much for your help, this is the result before I have modified the training and validation sets.
1, is it straightforward to look at the results of the validation after the training is done?

![1719118792856](https://github.com/PengchengShi1220/NexToU/assets/80230735/8d923b8f-2b34-4f54-a3f6-a693b8406c6d)

2, Do I still need to change the nnUNetPlans.json file to become the same patch as the paper to train BTCV?


"3d_fullres_nextou": {
    "inherits_from": "3d_fullres",
    "patch_size": [
        48,
        192,
        192
    ]
}
3, I see that BTCV is nnformer like validating 8 categories, what I am getting now should be all of them, may I ask where to go to change the settings I only need to validate eight categories. Is it the dataset.json file? Please help, thanks a lot!
![image](https://github.com/PengchengShi1220/NexToU/assets/80230735/7f0a0a55-167a-4b02-9c3c-f8d68c84eeff)

**Comment by PengchengShi1220** [comment]
Hi,
1. Validation Results: The Mean Validation Dice on the validation set is the average of 13 categories. Detailed results can be found in the `fold_0` directory under the `validation` folder. This folder contains the predictions on the validation set generated by `checkpoint_final.pth`. Additionally, the `summary.json` file includes the average metrics for each category across all validation cases, as well as metrics for each category per case. If you need to compute distance-based metrics such as HD 95 or NSD, please refer to the following resources:
   - HD 95: [NexToU inference](https://github.com/PengchengShi1220/NexToU/tree/NexToU_nnunetv2/inference)
   - NSD: [Google DeepMind](https://github.com/google-deepmind/surface-distance)
   
   In the future, I will update more validation metric calculation codes (NSD, clDice, and Betti number error).

2. Patch Size for BTCV:
   For BTCV, you can use a patch size of `[48, 192, 192]` consistent with nnUNet, which allows you to use `3d_fullres` directly. Alternatively, you can use `[64, 160, 160]`, ensuring each axis is a multiple of 32, which is necessary for NexToU.

3. Calculating Average Dice:
   As mentioned in point 1, you can get the case average Dice for each of the 13 categories from the summary.json file. You can then extract the Dice scores for the 8 categories you need and calculate their average.  Additionally, you can refer to the [NexToU inference](https://github.com/PengchengShi1220/NexToU/tree/NexToU_nnunetv2/inference) for further calculations. For results of NexToU_nnunetv1, please see [Google Drive](https://drive.google.com/drive/folders/1twMg6RSVIsxs6CsNirYpY1b5ihS0CRW_?usp=sharing).

If you have any further questions or need additional assistance, please feel free to reach out.

**Comment by qiaoqiangPro** [comment]
Hi, thank you very much for your answer.
1、I would also like to make sure that I should use the average in summary to compare with your paper, not the average here in inference_synapse.py, here it is just to get the results for each category in the paper table.
![image](https://github.com/PengchengShi1220/NexToU/assets/80230735/cbc69ae6-a876-4e91-ac73-c6db5695f88b)

![image](https://github.com/PengchengShi1220/NexToU/assets/80230735/15da86c8-3dfb-48df-a1a7-dab916f6643e)
2、For BTCV patch sizes I use [48, 192, 192] patch sizes and can also train NexToU. however it will report an error at first, but can be trained normally.
![image](https://github.com/PengchengShi1220/NexToU/assets/80230735/8a794df0-a0ad-4ec1-80f9-db8f37332fd3)

**Comment by PengchengShi1220** [comment]
Hi,
1. The average Dice values for the 8 categories in inference_synapse.py are indeed the same as those in summary.json. You can calculate and compare them to confirm this. 

2. I have encountered the warning you mentioned before as well. It is not caused by incorrect patch sizes, as incorrect patch sizes would result in an immediate error, preventing training. This warning is likely due to the torch.onnx or the version of pytorch and does not affect model training and inference.

**Comment by qiaoqiangPro** [comment]
Thank you very much for your patience and multiple replies, I still have questions about the above:
1, I verified the prompt result dice (Figure 1) for training completion, and the average dice in summary.json in the validation file (Figure 2). As well as the results run using the inference_synapse.py you provided (Figure 3). The first two have the same average dice, while the last one only calculates 8 categories, is this view okay?
I would like to enquire if the average of the thesis table is the result of 8 categories or 13 categories. If it's 13 categories then you have to calculate it in the same way as figure 1 and 2, if it's 8 categories then you need to use inference_synapse.py to calculate it. I would like to make sure from you.

<img width="478" alt="image" src="https://github.com/PengchengShi1220/NexToU/assets/80230735/aacdb70d-3c23-4428-8e3b-0de4e16cf935">
**fig1**
<img width="731" alt="image" src="https://github.com/PengchengShi1220/NexToU/assets/80230735/52fe509c-8ceb-4d48-8295-ff1fa743ecde">
**fig2**
<img width="390" alt="image" src="https://github.com/PengchengShi1220/NexToU/assets/80230735/fcabb59a-108b-4f6a-a0a7-abb22063db4a">
**fig3**
2, Didn't the paper mention that the patch of BTCV is [48,192,192]. I am following this and it works fine. Just have the error reported at the very beginning, which you have answered. However, you mentioned that it must be a multiple of 32, and 48 does not satisfy this condition, so I still need to confirm this with you.

**Comment by PengchengShi1220** [comment]
Thank you for your patience and for raising these questions.

1. **Dice Score Calculation**:
    - The value of 0.8506 shown in your figure is the average Dice score across 13 categories. The `inference_synapse.py` script uses the nnFormer code for calculations, while the average value in the paper is derived from 8 categories.
    - Each category’s corresponding Dice score calculated using `inference_synapse.py` and nnUNet should match the values in `summary.json`. You can verify this by checking each category individually in `summary.json`.

2. **Patch Size for BTCV Dataset**:
    - For the NexToU model, the patches in most datasets need to be multiples of 32 along each dimension. However, BTCV is an exception, where patch sizes can be [48, 192, 192]. This exception is related to the nnUNet's "pool_op_kernel_sizes" configuration for the BTCV dataset:
    ```
    [
        [1, 1, 1],
        [1, 2, 2],
        [2, 2, 2],
        [2, 2, 2],
        [2, 2, 2],
        [1, 2, 2]
    ]
    ```
    - The first pooling operation with [1, 2, 2] impacts the encoding dimensions. This allows BTCV to use [48, 192, 192] or [64, 160, 160] patches. Note that the NexToU v1.2 version included residual connections in the shallow stage.
    - I will update the new paper version, along with the new proposed loss functions and new datasets. The network structure code will similar to v1.2.

**Comment by qiaoqiangPro** [comment]
I've probably got it, thanks for your patience. All the best, I hope it gets better and better, and I look forward to your better and more exciting work, which I will continue to follow.

**Comment by qiaoqiangPro** [comment]
hello, Dear Shi:
I want to calculate the inference speed for different models but don't know how to calculate the inference speed under cpu and gpu in nnunet framework, can you provide help?

**Comment by PengchengShi1220** [comment]
Hello,

You can refer to the instructions provided in [nnunetv2_inference](https://github.com/MIC-DKFZ/nnUNet/tree/master/nnunetv2/inference), specifically the section "Predicting a single npy array," which can be used to infer a single image at a time. To use GPU for inference, set `perform_everything_on_device=True`. For CPU inference, set `perform_everything_on_device=False`. Here's an example code snippet:

```python
from nnunetv2.paths import nnUNet_results, nnUNet_raw
import torch
from batchgenerators.utilities.file_and_folder_operations import join
from nnunetv2.inference.predict_from_raw_data import nnUNetPredictor

# Instantiate the nnUNetPredictor
predictor = nnUNetPredictor(
    tile_step_size=0.5,
    use_gaussian=True,
    use_mirroring=True,
    perform_everything_on_device=True,  # Set this to False for CPU inference
    device=torch.device('cuda', 0),     # Use 'cpu' for CPU inference
    verbose=False,
    verbose_preprocessing=False,
    allow_tqdm=True
)

# Initialize the network architecture and load the checkpoint
predictor.initialize_from_trained_model_folder(
    join(nnUNet_results, 'Dataset003_Liver/nnUNetTrainer__nnUNetPlans__3d_fullres'),
    use_folds=(0,),
    checkpoint_name='checkpoint_final.pth',
)

# Predict a single numpy array (SimpleITKIO)
img, props = SimpleITKIO().read_images([join(nnUNet_raw, 'Dataset003_Liver/imagesTr/liver_63_0000.nii.gz')])
ret = predictor.predict_single_npy_array(img, props, None, None, False)
```

This code will help you perform inference on either a GPU or CPU within the nnU-Net framework. Let me know if you need any further assistance.

**Comment by qiaoqiangPro** [comment]
Hello dear shi.
Thank you very much for your patient answer and help, I have understood how the nnunetv2 framework uses cpu inference, which is very convenient. But it seems that the nnunetv1 code can't set the device directly, do you know how to use cpu reasoning in the model inference phase and measure the inference time?
